### DISCUSSION TITLE
Choosing few shots only when gold references/labels are available, possible in DSPy?

### DISCUSSION THREAD

**Discussion by brando90** [stanfordnlp/dspy#1493] [Q&A]
I have many gold reference **target/label**, but inputs in my case don't exist. Can DSPy do this? I want the few shots to be chosen at least based only on the right output/target distribution.

---

[Docs say](https://dspy-docs.vercel.app/docs/building-blocks/data):

> In DSPy, the Example objects have a with_inputs() method, which can mark specific fields as inputs. (The rest are just metadata or labels.)

but how does it know what is the label then?

Toy example

```python
import dspy

# Create an Example object
# qa_pair = dspy.Example(question="This is a question?", answer="This is an answer.")
qa_pair = dspy.Example(answer="This is an answer.")

# Specify the input field, leaving the answer as the output (label)
# qa_pair_with_input = qa_pair.with_labels("answer")

# print(qa_pair_with_input)
print(qa_pair)

#%%
# from dspy.datasets import HotPotQA

# # Load the dataset
# dataset = HotPotQA(train_seed=1, train_size=20, eval_seed=2023, dev_size=50, test_size=0)

# # Tell DSPy that the 'question' field is the input and the 'answer' field is the label
# trainset = [x.with_inputs('question').with_labels('answer') for x in dataset.train]
# devset = [x.with_inputs('question').with_labels('answer') for x in dataset.dev]

# len(trainset), len(devset)
```

**Comment by GioiaZheng** [comment]
DSPy does not have a separate `with_labels()` method. In the current `Example` implementation, `with_inputs(...)` marks only the input keys; `example.labels()` is then computed as "all fields that are not inputs."

So for the usual case:

```python
ex = dspy.Example(question="This is a question?", answer="This is an answer.").with_inputs("question")

ex.inputs().toDict()  # {"question": "..."}
ex.labels().toDict()  # {"answer": "..."}
```

If you create `dspy.Example(answer="...")` and never call `with_inputs(...)`, then `inputs()`/`labels()` are not defined and optimizers that call `example.inputs()` can fail. If you call `.with_inputs()` with no keys, then `answer` becomes a label, but that only makes sense for a signature with no input fields. It will not work as a normal `question -> answer` training example, because the demo does not contain the input field required by the signature.

For your target-distribution goal, the built-in few-shot optimizers do not currently do label-stratified demo selection by themselves:

- `LabeledFewShot` samples whole examples from the trainset.
- `BootstrapFewShot` runs the teacher on `example.inputs()` and filters through the metric.
- `KNNFewShot` retrieves demos by the current input, not by the label.

The safe pattern is to stratify outside DSPy, then pass selected full examples, with both inputs and labels, to the optimizer:

```python
selected = [
    dspy.Example(question=q, answer=a).with_inputs("question")
    for q, a in balanced_examples
]

compiled = dspy.LabeledFewShot(k=len(selected)).compile(
    program,
    trainset=selected,
    sample=False,
)
```

If your inputs truly do not exist, DSPy can still store those target values as fields on `Example`, but it cannot optimize or choose useful input-conditioned demonstrations for a normal task from labels alone. You would need either an input column, a zero-input signature, or a custom sampler/optimizer that treats the target distribution as metadata.
